rfd/0103-application-access-web-ui-auth-flow.md

```diff
@@ -22,53 +21,44 @@ the auth service. This application session has two values (a name and a bearer t
-as follows:
+as follows (using debug dumper app as example):
 
 ```mermaid
 sequenceDiagram
-    participant User
-    participant Proxy UI
-    participant Application
+    participant Browser
+    participant Proxy
+    participant Target App
     participant Auth Service
-    User->>Proxy UI: Launch Application
-    Proxy UI->>Auth Service: Create App Session /v1/webapi/sessions/app
-    Auth Service->>Proxy UI: Returns two cookie values (app session & subject)
-    Proxy UI->>Application: CORS preflight request to appurl.com/x-teleport-auth
-    Application->>Proxy UI: Matches Origin against public proxy addresses & allows CORS if a match
-    Proxy UI->>Application: POST appurl.com/x-teleport-auth with the cookie values in the header
-    Application->>Proxy UI: Sets the cookies on appurl.com if they are valid
-    Proxy UI->>User: User is logged in, redirect to appurl.com
-```
-
-### CORS
-
-As we need to enable cross-origin requests from the proxy URL to the application URL, we add middleware
-before `/x-teleport-auth` to check the `Origin` header (this cannot be changed via a `fetch` request).
-
-We check the origin against the public addresses of the proxy, and if one is a match, we allow a cross-origin
-request.
-
-The rules around the cross-origin request are strict and the absolute minimum needed:
+    alt Loading target app outside of web UI
+    Note left of Browser: User copy pastes target app URL into browser <br/>https://dumper.localhost:3080
+    Browser->>Proxy: Redirect to web launcher, preserving the requested path and query params <br/>In subsequent requests both the path and query gets preserved as path<br/>REDIRECT /web/launch/dumper.localhost?path=<requested-path>&query=<requested-query>
+    Note left of Browser: Redirected to app launcher
+    else Loading target app within the web UI (using the web app launcher)
+    Note left of Browser: At web UI, user clicks on `launch` button for the target app
+    Note left of Browser: Web UI routes to app launcher <br/>/web/launch/dumper.localhost/cluster-name/dumper.localhost<br/>route format: /web/launch/:fqdn/:clusterId?/:publicAddr?
+    end
+    Note right of Browser: App launcher navigates to target app <br/>https://dumper.localhost:3080/x-teleport-auth
+    Browser->>Target App: Navigating to target app start's the auth exchange <br/>GET /dumper.localhost/x-teleport-auth
+    Target App->>Browser: Create state token, set the token value in a cookie, and as a query param in the redirect URL back to web launcher<br/>REDIRECT /web/launch/dumper.localhost/cluster-name/dumper.localhost?state=<token>
+    Note left of Browser: Redirected back to app launcher
+    Browser->>Proxy: Create App Session <br>POST /v1/webapi/sessions/app with body:<br>{fqdn: dumper.localhost, plubic_addr: dumper.localhost, cluster_name: cluster-name}
+    Proxy->>Auth Service: Create App Session grpc.CreateAppSession
+    Auth Service->>Proxy: Returns two cookie values (app session & subject)

```

Comment:
Why two cookies? Do we need them both?

Reply:
Yes, when we create an app session, we create a web app session and a bearer token. We set one cookie with the id of app session, and we set another cookie with the value of the bearer token. This is how we authenticate request going to app service